### ISSUE TITLE
Public endpoint with per-country diagnosis keys for research and transparency

### ISSUE THREAD

**Issue by pvieito** [eu-federation-gateway-service/efgs-federation-gateway#209]
The EFGS project should provide a public endpoint so that once countries are sharing the diagnosis keys researchers can query the EFGS for the keys of each country.

For example, I have been working on [a project](https://github.com/pvieito/Radar-STATS) to monitor the Spanish Exposure Notification app “Radar COVID”, but once TEKs from other countries start appearing [mixed on the server with Spanish TEKs we won't be able to correctly monitor the performance of the app](https://github.com/RadarCOVID/radar-covid-ios/issues/20#issuecomment-701033756).

It would be great to have a public UE-wide API to download diagnosis keys of each country, is there any plan for that?

**Comment by pvieito** [comment]
Some people with great projects that may be also interested on something like this:
- @mh- ([corona-warn-companion-android](https://github.com/mh-/corona-warn-companion-android))
- @micb25 ([dka](https://github.com/micb25/dka))
- @sftcd ([tek_transparency](https://github.com/sftcd/tek_transparency))

**Comment by mikehelmick** [comment]
I would recommend against this at the EFGS level. The official recommendation from Google on export files is to pad them out so that the actual valid TEKs can't be distinguished from generated data. Since the EFGS only operates on actual TEKs, revealing them publicly here would render the padding ineffective for its intended purpose.

**Comment by mh-** [comment]
@mikehelmick 
Padding has been completely ineffective ever since it has been introduced, for the following reason:

Padding is only required when there are very few users within one country reporting their DKs within the server’s minimum grouping period (e.g. hour or day). 
In this situation, an adversary would connect DKs over several days (which is what padding is supposed to prevent) simply based on the **geographical positions where their RPIs were recorded**. If there is only one DK reporter in a small village across several days, they can assume it’s the same person. 
Additionally padded random DKs, to which no RPIs have been broadcast in the wild, will not be of any interest whatsoever to an adversary, it’s just noise floor in the system. 
A noise floor that is eating up CPU time and download capacity, and causing corresponding CO2 emissions, however.

I really wonder if you see any additional use for padding, that I missed.

**Comment by pvieito** [comment]
Not only that, but if the EFGS only operates on real TEKs it would be trivial to download them from two different country backends to identify common TEKs as real TEKs. If any padding is applied, it should be done before they are uploaded to the EFGS so all countries receive the same padding TEKs.

**Comment by sftcd** [comment]
On 16/10/2020 07:53, Pedro José Pereira Vieito wrote:
> Not only that, but if the EFGS only operates on real TEKs it would be
> trivial to download them from two different country backends to
> identify common TEKs as real TEKs. If any padding is applied, it
> should be done before they are uploaded to the EFGS so all countries
> receive the same padding TEKs.
> 

FWIW, I fail to see any utility at all in padding.
If one has no RPIs, it doesn't matter. If one has
some RPIs, it still doesn't matter.

And in practice we were able to identify the fakes in
Austria (not transmitted often enough) and can still
estimate the number of uploads in Poland after they
introduced padding (a tiny number of changes all of
around 1000 TEKs at a time). I don't see a threat
model where padding makes some sense here. (Unless
independent measurement is considered a threat;-)

S.

**Comment by pvieito** [comment]
> FWIW, I fail to see any utility at all in padding. If one has no RPIs, it doesn't matter. If one has some RPIs, it still doesn't matter.

Completely agree!

> And in practice we were able to identify the fakes in Austria (not transmitted often enough) and can still estimate the number of uploads in Poland after they introduced padding (a tiny number of changes all of around 1000 TEKs at a time). I don't see a threat model where padding makes some sense here. (Unless independent measurement is considered a threat;-)

Yep! Another example is Malta, they apply a 100 TEK padding but the padding TEKs change each day so it is trivial to detect the real ones as the only ones that do not disappear 😂

**Comment by mikehelmick** [comment]
The objective of the padding is to obscure chains of TEKs so that those belonging to an individual device are unlikely to be tied together. The v1.5 fields have made this proposition more difficult since there is more metadata that varies from record to record.

If another source choses to make the TEKs for a country available wholesale, then the padding is rendered entirely ineffective.

We may take the stance of publishing our padding as-is to EFGS so that no difference can be derived from cross referencing the sources.

@pvieito - In the case you mentioned. It sounds like valid TEKs are being re-published instead of incremental exports, which is the recommendation.

**Comment by mh-** [comment]
@mikehelmick You seem to assume that an adversary might be interested in tying chains of TEKs together - however, adversaries are not interested in TEKs as such. They would be interested in tying RPIs together.
And they can do that based on the geographical location where they recorded the RPIs.
That's why **padding serves no purpose at all**, except maybe to create a false sense of privacy.

**Comment by mikehelmick** [comment]
The padding is not to defend against someone who is collecting RPIs and geotagging them.

**Comment by mh-** [comment]
So what is it trying to defend against?

**Comment by pvieito** [comment]
Hmm… the only advantage I see to padding is in the case of adversary with huge dataset of geolocated RPIs, in that case if he had a TEK chain of 14 days he would be able to track a unique device for 14 days instead of 1. This is a very improbable case for countries which have an Exposure Notification app with big penetration, but it would be possible (ie. if the user shares the diagnose at night).

**Comment by mikehelmick** [comment]
@mh- I send you an email at your gmail address.

@pvieito - correct, in our model we phase out padding if there is enough real data.

**Comment by pvieito** [comment]
> @pvieito - correct, in our model we phase out padding if there is enough real data.

Great! But going back to the main point of this issue, this padding should be applied on the ingestion phase, so I don't think there should be any problem at exposing the EFGS per-country TEKs endpoint publicly.

**Comment by sftcd** [comment]
On 16/10/2020 19:47, Pedro José Pereira Vieito wrote:
> Hmm… the only advantage I see to padding is in the case of adversary
> with huge dataset of geolocated RPIs, in that case if he had a TEK
> chain of 14 days he would be able to track a unique device for 14
> days instead of 1. 

How does adding fake TEKs help? If the real ones are there
they match the real RPIs. And the fake TEKs cannot. And you
cannot elide any of the real TEKs.

No - there's zero value in fake TEKs. (And it's not
"padding" according to any definition related to trafic
analysis with which I'm familiar.)

S.

**Comment by sftcd** [comment]
On 16/10/2020 19:50, Mike Helmick wrote:
> @mh- I send you an email at your gmail address.
> 
> @pvieito - correct, in our model we phase out padding if there is enough real data.

Where is that model?

Thanks,
S.

**Comment by pvieito** [comment]
> How does adding fake TEKs help? If the real ones are there they match the real RPIs. And the fake TEKs cannot. And you cannot elide any of the real TEKs.

But even if the adversary removes the “padding” and undetected TEKs using the RPI dataset he cannot know (but it may be able to presume it) that the 14 TEKs are from the same device (as the “padding” guarantees that may be from different chains).

**Comment by mh-** [comment]
> > How does adding fake TEKs help? If the real ones are there they match the real RPIs. And the fake TEKs cannot. And you cannot elide any of the real TEKs.
> 
> But even if the adversary removes the “padding” and undetected TEKs using the RPI dataset he cannot know (but it may be able to presume it) that the 14 TEKs are from the same device (as the “padding” guarantees that may be from different chains).

Even if there are only exactly 14 TEKs in one batch, the adversary cannot know with 100% certainty that they belong to the same user. They may as well belong to e.g. 3 users, who had their phones switched off at different times.

**Comment by pvieito** [comment]
@mh- Yes, that is also true but very improbable.

**Comment by mh-** [comment]
@pvieito exactly, it's very improbable. 
And if an adversary records RPIs where they find matching TEKs on multiple days in the same batch in a situation when only few user are reporting their diagnosis, it's also very improbable that those were two different individuals. Regardless of added padding noise.

**Comment by pvieito** [comment]
@mh- Totally agree! 

(To be clear: I am on the non-padding team here 😂)

**Comment by kbobrowski** [comment]
Let's assume someone is operating widely deployed RPI scanner network, and is recording signals from person A and person B on day 1 (and can tell to which person RPIs belong on that particular day using other means). Both A and B spend night together, and then part ways. Adversary continues scanning on day 2. Both A and B are diagnosed and share their TEKs at the same time.

If adversary can chain TEKs together, then he knows with certainty where person A went on day 2. Otherwise he knows it with 50% probability (since other possible place is where B went).

**Comment by mh-** [comment]
@kbobrowski
If A and B share their TEKs at the same time, both TEK chains are in the same package in parallel - how can adversary chain them?

**Comment by kbobrowski** [comment]
If TEKs are being sorted then they will have to land in different packages for the adversary to be able to chain them. A and B may actually upload it on different days (or if the packaging is done more often than every 24 hours), and then they are automatically chained regardless of sorting or padding, just based on scanned RPI data (if the data is fine-grained enough). Never understood benefit of padding with random data, for all purposes this seems to be just noise

**Comment by f11h** [comment]
Feature is out of scope.
